### PR TITLE
feat: Network view should group pods into higher level workload (#5468)

### LINKED ISSUE #5468: Network view should group pods into higher level workload
# Summary

The network view does not look good when you have a large application with many pods. 

image

Most of the time, users don't care how traffic is routed to individual pods, but rather to some higher level workload that created that pod (e.g. ReplicaSet).

# Proposal

Instead of showing lines to many pods, we should connect the service to the parent of the pod. Possibly leveraging what we did in the pod view:

image

### PR DESCRIPTION
**Pull request by keithchong**
Signed-off-by: Keith Chong <kykchong@redhat.com>

Fixes https://github.com/argoproj/argo-cd/issues/5468.  This Pod Group applies to both the Details Network view and the Tree view. The Pod Group is made available to the tree-based views so that it can be integrated with the Tree Expansion support. eg. the Pod Group can be the collapsed state for the higher level workload, and when you expand it, it will expand out the Pods.

Notes:

At a high level, the general solution involves flagging all potential Pod Groups (that is, “parents” of pods in the graph views). It piggy backs on the existing logic (loops) when gathering up the graph nodes. Once a pod is found, it can be a graph node, or not, and it depends on the property setting showCompactNodes, which I’m using to turn the feature on and off. The pod is then added to the pod group and it continues on with the next iteration of the loop.

I initially wanted to make the pod group look like the pod group from the Pods view.  But it was better to try to make it look like the existing nodes in the tree and network views so that the animation transition looks smoother.
- See sample videos and screenshots below
- Changing to reparent to RS can come later.
- No QuickStart Buttons
- In details tree view, Grouped Nodes feature supersedes this (unless we want a separate toolbar button for this)
- For the Network View, I marked the Pod Group type as parentResource since that seems like the suitable fit. I could not create a new type, otherwise it will show up in the Pods View filter.
- Moved PodView’s deletePopup to utils.tsx so it can be reused
- Pod Height and layout adjusts accordingly when more rows are added to accommodate more pods

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by keithchong** [comment]
https://user-images.githubusercontent.com/7726022/161667068-242a2d8e-08a8-4db9-bac8-c6949948c68f.mov

https://user-images.githubusercontent.com/7726022/161667106-83fc6887-ebe8-489d-87c8-be76cf47207c.mov

**Comment by keithchong** [comment]
https://user-images.githubusercontent.com/7726022/161667844-5aa88941-39c9-4a0e-bb34-3ca7f0e11945.mov

https://user-images.githubusercontent.com/7726022/161667866-c67cb90a-2d6d-441a-b4f5-82082e389814.mov

**Comment by keithchong** [comment]
https://user-images.githubusercontent.com/7726022/161669311-14b95861-0fb0-4dd5-8b3d-4b505ef7e9a0.mov

https://user-images.githubusercontent.com/7726022/161669287-37bb3a91-295b-4569-a887-a2b0d0a540e2.mov

**Comment by keithchong** [comment]
canary
cert-manager
prometheus-operator

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 1]
```diff
@@ -307,6 +324,244 @@ export const describeNode = (node: ResourceTreeNode) => {
...
+                                                }
+                                            }
+                                        ]}
+                                    />
+                                ))}
+                            </div>
+                            <div className='application-resource-tree__node--lower-section__pod-group__label'>PODS</div>
+                        </div>
+                    </div>
+                    // </div>
```
Comment (ciiay) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:558]: Unnecessary comment out.

**Review by ciiay — APPROVED** [review]
Looks great!! Since the grouped pod view looks very similar to the Pods view, is there a simple way to make it partially reusable component? Just curious.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 1 above)
Comment (rbreeze) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:558]: +1

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 2]
```diff
@@ -513,8 +768,17 @@ export const ApplicationResourceTree = (props: ApplicationResourceTreeProps) =>
             findNetworkTargets(networkNodes, parent.networkingInfo).forEach(child => {
                 const children = childrenByParentKey.get(treeNodeKey(parent)) || [];
                 hasParents.add(treeNodeKey(child));
-                children.push(child);
-                childrenByParentKey.set(treeNodeKey(parent), children);
+                if (child.kind !== 'Pod') {
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:771]: I think we can avoid repeating these lines:
```
children.push(child);
childrenByParentKey.set(treeNodeKey(parent), children);
```

By doing something like:
```
if (!props.showCompactNodes || child.kind !== 'Pod') {
    // those lines
} else {
    processPodGroup(parent, child, props);
}

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx [hunk 3]
```diff
@@ -589,8 +853,18 @@ export const ApplicationResourceTree = (props: ApplicationResourceTreeProps) =>
                 }
                 node.parentRefs.forEach(parent => {
                     const children = childrenByParentKey.get(treeNodeKey(parent)) || [];
-                    children.push(node);
-                    childrenByParentKey.set(treeNodeKey(parent), children);
+                    if (node.kind !== 'Pod') {
```
Comment (rbreeze) [inline ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:856]: same comment here as above

**Review by rbreeze — CHANGES_REQUESTED** [review]
A couple of small things, otherwise looks good. I also agree with @ciiay 's comment about potentially factoring out the inner pod component into a shared component between this and pod view. That can perhaps be done later though.

Thanks!

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 1 above)
Comment (keithchong) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:558]: Acknowledged.  Done.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 2 above)
Comment (keithchong) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:771]: Thanks Remington.  Agreed.

File: ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx (hunk 3 above)
Comment (keithchong) [reply ui/src/app/applications/components/application-resource-tree/application-resource-tree.tsx:856]: Agreed.

**Comment by keithchong** [comment]
I opened https://github.com/argoproj/argo-cd/issues/9338 to track the two review comments to pull out the pod group into a common component so that both the Tree view and the Pods view can use.

**Review by rbreeze — APPROVED** [review]
LGTM, thanks @keithchong!

**Comment by keithchong** [comment]
Thanks @rbreeze

**Pull request merged**